File: solutions/scala/problem-00022.scala
/*
[minhuang@mocha:scala]$ time scala problem-00022.scala 
871198282

real  0m2.015s
user  0m0.988s
sys 0m0.084s
*/
import scala.io.Source
import scala.Math._

val scores = Map() ++ (('A' to 'Z').toList zip (1 to 26).toList)

def score(name: String) = name.map(c => scores(c)).reduceLeft(_ + _)

val data = Source.fromFile("../../data/problem-00022.txt")
                 .getLines
                 .toList
                 .first
                 .replaceAll("\"", "")
                 .split(",")
                 .toList

val x = (data.indices zip data.sort(_ < _)).map(i => {
  (i._1 + 1) * score(i._2)
}).reduceLeft(_ + _)
println(x)

File: solutions/ruby/problem-00022.rb
#!/usr/bin/env ruby
# [minhuang@mocha:ruby]$ time ruby problem-00022.rb 
# 871198282
#
# real  0m1.942s
# user  0m1.928s
# sys 0m0.008s
def score(name)
  scores = Hash[('A'..'Z').zip(1..26)]
  name.chars.map { |c| scores[c] }.inject(&:+)
end

file = File.new("../../data/problem-00022.txt")
data = file.gets.gsub("\"", "").split(",") # All one line in file
file.close

result = data.each_index.
              zip(data.sort).
              map { |t| (t.first+1) * score(t.last) }.
              inject(&:+)
puts result
